### PR TITLE
[Dont merge yet] - Grid search

### PR DESCRIPTION
**Pull request by SebastianRehfeldt**
* added class which calls all feature selection functions with different parameters
* class collects all information and writes to csv file
* changed some parameters of feature selection methods
* changes to EA  (gets k and a new pre_selection number)

##TODO##
* check if feature selection methods are called correctly in app.py and all other files

### CODE REVIEW

File: utils/EA/config.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ### General ###
 maximization = True
 runs = 1
-generations = 50
+generations = 5
```
Comment (lukasjf) [inline utils/EA/config.py]: Is this temporary? I recall more iterations to have helped?

File: gridSearch.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from validation.GridSearch import GridSearch
+
+print("Imported modules", flush=True)
+
+dataLoader = DataLoader("dataset4")
+print("data loaded", flush=True)
+
+healthy = dataLoader.getData(["healthy"], ["THCA","LUAD"])
+sick = dataLoader.getData(["sick"], ["THCA","LUAD"])
+data = dataLoader.getData(["sick", "healthy"], ["THCA","LUAD"])
```
Comment (lukasjf) [inline gridSearch.py:15]: Let's not forget the other datasets later ;)

File: grid_search_one_vs_rest.csv [hunk 3]
```diff
@@ -0,0 +1,11 @@
...

+2,THCA,norm,3,0.622,0.899,0.83,"[4126, 5632, 4187]"

+3,LUAD,ea,3,0.625,0.959,13.14,"[4126, 14530, 17522]"

+3,THCA,ea,3,0.679,0.94,13.14,"[14530, 4101, 12336]"

+4,LUAD,sfs,3,0.643,0.955,16.24,"[4126, 7818, 26238]"

+4,THCA,sfs,3,0.672,0.927,16.24,"[4126, 15249, 16675]"
```
Comment (lukasjf) [inline grid_search_one_vs_rest.csv]: both tables look fine

File: utils/DimensionalityReducer.py [hunk 4]
```diff
@@ -100,19 +100,19 @@ def getNormalizedFeaturesE(self, sick, healthy, k, n, m, returnMultipleSets = Fa
...
         mutation = binary_mutation
 
         sick_reduced = Expressions(sick.expressions[:, selected_genes], sick.labels)
         healthy_reduced = Expressions(healthy.expressions[:, selected_genes], healthy.labels)
 
-        fitness_function = fitness_module.fitness(sick_reduced, healthy_reduced, fitness)
+        fitness_function = fitness_module.fitness(sick_reduced, healthy_reduced, fitness, k, true_label)
 
-        best, sets, _, _ = ea_for_plot(c, c.chromo_size, fitness_function, crossover, mutation)
+        best, sets, _, _ = ea_for_plot(c, m, k, fitness_function, crossover, mutation)
```
Comment (lukasjf) [inline utils/DimensionalityReducer.py:113]: m replaced chromo_size

File: utils/EA/config.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
...
 population_size = 100
 chromo_size = 100
 prob_mutation = 1/generations
```
Comment (lukasjf) [inline utils/EA/config.py]: shouldn't we do 1/chromo_size

with generations  = 5 I suppose this to be very random

File: utils/EA/population.py [hunk 6]
```diff
@@ -3,14 +3,14 @@
...
+def generate_population(size_cromo, k, size_pop = population_size):
+    return [ (generate_indiv(size_cromo, k), 0) for i in range(size_pop)]
 
-def generate_indiv(size_cromo):
+def generate_indiv(size_cromo, k):
     # random initialization
     indiv = [0 for i in range(size_cromo)]
 
-    for i in range(10):
+    for i in range(k):
```
Comment (lukasjf) [inline utils/EA/population.py:13]: maybe for later: start with different sizes; avoid collisions

File: validation/GridSearch.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+
+    ### Utils ###
+    def get_table_one_vs_rest(self):
+        if not self.table_one_vs_rest is None:
+            return self.table_one_vs_rest
+
+        table = []
+        table.append(["ID", "Type", "Method", "K", "Fitness_score", "Sick_F1", "Time", "Features"])
+        table.extend(self.get_one_against_rest_results())
```
Comment (lukasjf) [inline validation/GridSearch.py:162]: I find it a bit odd, to hide the main call in the lazy getter

File: validation/GridSearch.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        table.append(["Method", "K", "Statistic", "Normalization", "Exclude", "Preselect", "Fitness_method", "Fitness_score", "Sick_F1", "Time", "Features"])
+        table.extend(self.get_basic_results())
+        table.extend(self.get_normalized_results())
+        table.extend(self.get_combined_results())
+
+        self.table_all_at_once = table
+        return table
+
+    def get_result_dict_one_vs_rest(self, id, cancer_type, method, k, metrics, time, features):
+        fitness = float(str(metrics["fitness"]["combinedFitness"])[0:5])
```
Comment (lukasjf) [inline validation/GridSearch.py]: note for me to again look at the structure of the metric parameter tomorrow
